File: src/js/tests/articulation.test.ts
import { expect, test } from 'vitest';
import { Articulation } from '@model';

test('defaultArticulation', () => {
  const a = new Articulation();
  expect(a).toBeInstanceOf(Articulation);
  expect(a.name).toEqual('pluck');
  expect(a.stroke).toEqual(undefined);
  expect(a.hindi).toEqual(undefined);
  expect(a.ipa).toEqual(undefined);
  expect(a.engTrans).toEqual(undefined);
});

test('Articulation fromJSON', () => {
  const obj = { name: 'pluck', stroke: 'd', hindi: 'द', ipa: 'd̪', engTrans: 'da', strokeNickname: 'da' };
  const a = Articulation.fromJSON(obj);
  expect(a.stroke).toBe('d');
  expect(a.strokeNickname).toBe('da');
});

test('strokeNickname defaults to da for d stroke', () => {
  const a = new Articulation({ stroke: 'd' });
  expect(a.strokeNickname).toBe('da');
  expect(a.name).toBe('pluck');
  expect(a.stroke).toBe('d');
  expect(a.hindi).toBeUndefined();
  expect(a.ipa).toBeUndefined();
  expect(a.engTrans).toBeUndefined();
});

test('strokeNickname defaults to da for d stroke', () => {
  const a = new Articulation({ stroke: 'd' });
  expect(a.strokeNickname).toBe('da');
  expect(a.name).toBe('pluck');
  expect(a.stroke).toBe('d');
  expect(a.hindi).toBeUndefined();
  expect(a.ipa).toBeUndefined();
  expect(a.engTrans).toBeUndefined();
});

test('stroke r sets strokeNickname', () => {
  const a = new Articulation({ stroke: 'r' });
  expect(a.strokeNickname).toBe('ra');
});
  
test('stroke r via fromJSON sets strokeNickname', () => {
  const obj = { name: 'pluck', stroke: 'r' };
  const a = Articulation.fromJSON(obj);
  expect(a.strokeNickname).toBe('ra');
});

File: python/api/tests/articulation_test.py
from python.api.classes.articulation import Articulation

def test_default_articulation():
    a = Articulation()
    assert isinstance(a, Articulation)
    assert a.name == 'pluck'
    assert a.stroke is None
    assert a.hindi is None
    assert a.ipa is None
    assert a.eng_trans is None


def test_articulation_from_json():
    obj = {
        'name': 'pluck',
        'stroke': 'd',
        'hindi': '\u0926',
        'ipa': 'd̪',
        'engTrans': 'da',
        'strokeNickname': 'da'
    }
    a = Articulation.from_json(obj)
    assert a.stroke == 'd'
    assert a.stroke_nickname == 'da'


def test_stroke_nickname_defaults_da():
    a = Articulation({'stroke': 'd'})
    assert a.stroke_nickname == 'da'
    assert a.name == 'pluck'
    assert a.stroke == 'd'
    assert a.hindi is None
    assert a.ipa is None
    assert a.eng_trans is None


def test_stroke_nickname_defaults_da_duplicate():
    a = Articulation({'stroke': 'd'})
    assert a.stroke_nickname == 'da'
    assert a.name == 'pluck'
    assert a.stroke == 'd'
    assert a.hindi is None
    assert a.ipa is None
    assert a.eng_trans is None


def test_stroke_r_sets_stroke_nickname():
    a = Articulation({'stroke': 'r'})
    assert a.stroke_nickname == 'ra'


def test_stroke_r_from_json_sets_stroke_nickname():
    obj = {'name': 'pluck', 'stroke': 'r'}
    a = Articulation.from_json(obj)
    assert a.stroke_nickname == 'ra'
